### DISCUSSION TITLE
Best way to set up an on-premises clone of vcpkg including bootstrapping?

### DISCUSSION THREAD

**Discussion by jdx-john** [microsoft/vcpkg#20025] [Q&A] [answered]
Our project security is very locked down meaning no internet access to development machines. I would like to have an on-premises local clone of vcpkg so developers can still get it using git as normal, and also so we can make custom changes. We'd download sources for ports onto a single machine (which has some internet access) and developers would point their vcpkg installations at this network location for pre-downloaded sources.

We can then periodically update our local repo and developers can pull from it then to ensure we keep current with versions, and everyone is using the same.

Cloning vcpkg is easy in DevOps but of course the bootstrapper hard-codes downloading a specific file from vcpkg-tool repo. 
I was thinking we can just edit the bootstrapper script to get a local file instead (which we can then periodically update on the central server). We could clone vcpkg-tool locally too but it seems a little unnecessary to me. This way developers still run vcpkg almost entirely as normal without needing internet access, but we retain everything on-premises.

Does all this seem reasonable? Have I missed any obvious improvements or problems with such an approach? Is there a simpler/better way to get the same result?

**Comment by autoantwort** [comment, marked as answer]
Seems reasonable. Instead of `--downloads-root=` we use [asset caching](https://github.com/microsoft/vcpkg/blob/master/docs/users/assetcaching.md), because not every body has set up the network drive, but in your situation it is reasonable.  
And instead of changing the bootstrap script we have committed the vcpkg.exe to the repo, so that you don't have to run the bootstrap script when we update the vcpkg submodule.

**Reply by jdx-john** [reply]
I've not tried asset caching, is this new? We've used binary caching before but it seemed tremendously fragile what would cause the cache to be considered incompatible.

I was trying to remember if just copying the `download` folder between machines was enough to share sources provided everyone is happy to rebuild.

Committing vcpkg.exe seems reasonable in the interests of simplicity too. Thanks.

**Reply by autoantwort** [reply]
Asset caching is relative new (since June). In comparison to binary caching it is not fragile, because it simply uses the SHA512 of the downloaded file as key.

Yeah copying the `download` folder is enough. We first committed the whole download folder to the internal vcpkg fork, but then the git repo grows very fast in size. But then happily asset caching was implemented.
